subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/TaskUpToDateIntegrationTest.groovy

```diff
@@ -101,6 +101,57 @@ class TaskUpToDateIntegrationTest extends AbstractIntegrationSpec {
         executedAndNotSkipped ":customTask"
     }
 
+    @Unroll
+    @Issue("https://github.com/gradle/gradle/issues/3073")
+    def "output files changed from #before to #after marks task up-to-date"() {
+        buildFile << """
+            class CustomTask extends DefaultTask {
+                @OutputFiles
+                List<Object> outputFiles
+            
+                @TaskAction
+                void doAction() {
+                    outputFiles.each { output ->
+                         def outputFile = output.call()
+                         if (outputFile != null) {
+                            outputFile.text = "task executed"
+                         }
+                    }
+                }
+            }
+            
+            def lazyProperty(String name) {
+                def value = project.findProperty(name)
+                def outputFile = value ? file(value) : null
+                return { -> outputFile }
+            }
+            
+            task customTask(type: CustomTask) {
+                outputFiles = [lazyProperty('output0'), lazyProperty('output1')]
+            }
+        """
+
+        when:
+        succeeds customTaskWithOutputs(before)
+        then:
+        executedAndNotSkipped ":customTask"
+
+        when:
+        succeeds customTaskWithOutputs(after)
+        then:
+        skipped ":customTask"
+
+        where:
+        before                | after
+        [null, 'outputFile']  | ['outputFile', null]
+        ['outputFile1', null] | ['outputFile1', 'outputFile2']
```

Comment:
I don't understand.

* `[null, 'outputFile']` --> `['outputFile', null]` should be up-to-date, so the test looks good, but:
* `['outputFile1', null]` --> `['outputFile1', 'outputFile2']` should **not** be up-to-date, no?

Reply:
That's what I was trying to explain before. For up-to-date checks of output properties we currently detect if
- The names of the output properties change (property added, removed, renamed)
- The actual file snapshot of the file collection for each output property is different. We compare the output file snapshot from the task history to the snapshot of the file collection which makes up the property. Here we ignore added files, since we still allow overlapping task outputs.

What we do not track is the set of declared output files per output file property.
In the example above, where you think it should not be up-to-date, we have one output property name: `outputFiles`, which is the same as in the task history. From the task history, we have a file snapshot from when we executed the task before. This snapshot is `[outputFile1]`. Now we take a snapshot of the `outputFiles` file collection for the up-to-date check. The result is `[outputFile1, outputFile2 (missing)]`. We are ignoring added snapshots, so we consider the snapshots the same. And so the task is up-to-date.

The declared output files would be `[outputFile1]` and `[outputFile1, outputFile2]` respectively. So if we would take those into account for change detection, then the task would not be up-to-date and execute.

For 4.0, we were still matching the best previous task execution. This compared the declared output files, under some circumstances, resulted in making the task out of date if they were different.

I think we have two different ways to go forward:
* Accept the current behavior, keeping the task up-to-date in some corner cases where the outputs are changed dynamically by the build script
* Make sure `OutputFilesTaskStateChanges` also tracks changes to the declared output files. I would do this per output property. It is not too hard to implement

WDYT?